File: src/diffler/helpers/filters.py
"""Repository filtering and sorting helpers."""

from __future__ import annotations

from typing import Callable


def _normalize(repo: dict) -> dict:
    """Ensure all expected keys exist with sensible defaults."""
    return {
        "name": repo.get("name") or "",
        "full_name": repo.get("full_name") or "",
        "description": repo.get("description") or "",
        "url": repo.get("url") or "",
        "stars": repo.get("stars") or 0,
        "forks": repo.get("forks") or 0,
        "language": repo.get("language"),
        "language_color": repo.get("language_color"),
        "is_fork": repo.get("is_fork", False),
        "is_archived": repo.get("is_archived", False),
    }


def filter_repos(
    repos: list[dict],
    *,
    language: str | None = None,
    min_stars: int | None = None,
    max_stars: int | None = None,
    exclude_forks: bool = False,
    exclude_archived: bool = False,
    search: str | None = None,
    sort_by: str = "stars",
    sort_desc: bool = True,
    limit: int | None = None,
) -> list[dict]:
    """Filter and sort a list of repository dicts.

    Args:
        repos: List of repository dicts from the GitHub API.
        language: Filter by primary language (case-insensitive).
        min_stars: Minimum star count (inclusive).
        max_stars: Maximum star count (inclusive).
        exclude_forks: Skip forked repositories.
        exclude_archived: Skip archived repositories.
        search: Substring to match against name or description.
        sort_by: Field to sort by — "stars", "forks", "name", "language".
        sort_desc: Sort descending (default) or ascending.
        limit: Maximum number of results to return.

    Returns:
        Filtered, sorted, and optionally limited list of repository dicts.
    """
    normalized = [_normalize(r) for r in repos]
    result = list(normalized)

    if exclude_forks:
        result = [r for r in result if not r["is_fork"]]

    if exclude_archived:
        result = [r for r in result if not r["is_archived"]]

    if language:
        lang_lower = language.lower()
        result = [
            r for r in result
            if r["language"] and r["language"].lower() == lang_lower
        ]

    if min_stars is not None:
        result = [r for r in result if r["stars"] >= min_stars]

    if max_stars is not None:
        result = [r for r in result if r["stars"] <= max_stars]

    if search:
        term = search.lower()
        result = [
            r for r in result
            if term in r["name"].lower()
            or term in (r["description"] or "").lower()
            or term in (r["full_name"] or "").lower()
        ]

    # Sorting
    sort_key: Callable
    if sort_by == "stars":
        sort_key = lambda r: r["stars"]
    elif sort_by == "forks":
        sort_key = lambda r: r["forks"]
    elif sort_by == "name":
        sort_key = lambda r: r["name"].lower()
    elif sort_by == "language":
        sort_key = lambda r: (r["language"] or "").lower()
    else:
        sort_key = lambda r: r["stars"]

    result = sorted(result, key=sort_key, reverse=sort_desc)

    if limit is not None:
        result = result[:limit]

    return result


def repos_by_language(repos: list[dict]) -> dict[str, list[dict]]:
    """Group repositories by their primary language.

    Args:
        repos: List of repository dicts.

    Returns:
        Dict mapping language name to list of repos.
    """
    groups: dict[str, list[dict]] = {}
    for repo in repos:
        lang = repo.get("language") or "Unknown"
        groups.setdefault(lang, []).append(_normalize(repo))
    # Sort each group by stars descending
    for lang in groups:
        groups[lang] = sorted(groups[lang], key=lambda r: r["stars"], reverse=True)
    return dict(sorted(groups.items()))


def language_breakdown(repos: list[dict]) -> list[dict]:
    """Return a summary of languages used across repositories.

    Args:
        repos: List of repository dicts.

    Returns:
        List of dicts with keys: language, count, total_stars, color.
    """
    stats: dict[str, dict] = {}
    for repo in repos:
        lang = repo.get("language") or "Unknown"
        color = repo.get("language_color")
        if lang not in stats:
            stats[lang] = {"language": lang, "count": 0, "total_stars": 0, "color": color}
        stats[lang]["count"] += 1
        stats[lang]["total_stars"] += repo.get("stars", 0)

    result = sorted(stats.values(), key=lambda s: s["total_stars"], reverse=True)
    return result

File: packages/diffler/src/helpers/filters.ts
interface Repo {
  name: string;
  full_name: string;
  description: string | null;
  url: string;
  stars: number;
  forks: number;
  language: string | null;
  language_color: string | null;
  is_fork: boolean;
  is_archived: boolean;
}

function normalize(repo: Record<string, unknown>): Repo {
  return {
    name: (repo.name as string) || "",
    full_name: (repo.full_name as string) || "",
    description: (repo.description as string) || null,
    url: (repo.url as string) || "",
    stars: (repo.stars as number) || 0,
    forks: (repo.forks as number) || 0,
    language: (repo.primary_language as string) || null,
    language_color: (repo.primary_language_color as string) || null,
    is_fork: (repo.is_fork as boolean) || false,
    is_archived: (repo.is_archived as boolean) || false,
  };
}

export function filterRepos(
  repos: Array<Record<string, unknown>>,
  options: {
    language?: string;
    min_stars?: number;
    max_stars?: number;
    exclude_forks?: boolean;
    exclude_archived?: boolean;
    search?: string;
    sort_by?: string;
    sort_desc?: boolean;
    limit?: number;
  } = {}
): Array<Record<string, unknown>> {
  const normalized = repos.map(normalize);
  let result = [...normalized];

  if (options.exclude_forks) {
    result = result.filter((r) => !r.is_fork);
  }
  if (options.exclude_archived) {
    result = result.filter((r) => !r.is_archived);
  }
  if (options.language) {
    const lang = options.language.toLowerCase();
    result = result.filter((r) => r.language && r.language.toLowerCase() === lang);
  }
  if (options.min_stars !== undefined) {
    result = result.filter((r) => r.stars >= options.min_stars!);
  }
  if (options.max_stars !== undefined) {
    result = result.filter((r) => r.stars <= options.max_stars!);
  }
  if (options.search) {
    const term = options.search.toLowerCase();
    result = result.filter(
      (r) =>
        r.name.toLowerCase().includes(term) ||
        (r.description || "").toLowerCase().includes(term) ||
        r.full_name.toLowerCase().includes(term)
    );
  }

  const sortBy = options.sort_by || "stars";
  const sortDesc = options.sort_desc !== false;

  const sortKey = (r: Repo): string | number => {
    if (sortBy === "stars") return r.stars;
    if (sortBy === "forks") return r.forks;
    if (sortBy === "name") return r.name.toLowerCase();
    if (sortBy === "language") return (r.language || "").toLowerCase();
    return r.stars;
  };

  result.sort((a, b) => {
    const av = sortKey(a);
    const bv = sortKey(b);
    if (av < bv) return sortDesc ? 1 : -1;
    if (av > bv) return sortDesc ? -1 : 1;
    return 0;
  });

  if (options.limit !== undefined) {
    result = result.slice(0, options.limit);
  }

  return result as unknown as Array<Record<string, unknown>>;
}

export function reposByLanguage(
  repos: Array<Record<string, unknown>>
): Record<string, Array<Record<string, unknown>>> {
  const groups: Record<string, Array<Record<string, unknown>>> = {};
  for (const repo of repos) {
    const lang = (repo.primary_language as string) || "Unknown";
    groups[lang] = groups[lang] || [];
    groups[lang].push(repo);
  }
  for (const lang of Object.keys(groups)) {
    groups[lang].sort((a, b) => ((b.stars as number) || 0) - ((a.stars as number) || 0));
  }
  return Object.fromEntries(Object.entries(groups).sort(([a], [b]) => a.localeCompare(b)));
}

export function languageBreakdown(repos: Array<Record<string, unknown>>): Array<Record<string, unknown>> {
  const stats: Record<string, { language: string; count: number; total_stars: number; color: string | null }> = {};
  for (const repo of repos) {
    const lang = (repo.primary_language as string) || "Unknown";
    const color = (repo.primary_language_color as string) || null;
    if (!stats[lang]) {
      stats[lang] = { language: lang, count: 0, total_stars: 0, color };
    }
    stats[lang].count += 1;
    stats[lang].total_stars += (repo.stars as number) || 0;
  }
  return Object.values(stats).sort((a, b) => b.total_stars - a.total_stars);
}
